src/biokbase/narrative/jobs/specmanager.py

```diff
@@ -1,40 +1,53 @@
+"""Code for managing app specs."""
 import json
+from typing import Any
 
 from IPython.display import HTML
 from jinja2 import Template
 
-import biokbase.narrative.clients as clients
+from biokbase.narrative import clients
 from biokbase.narrative.app_util import app_param, app_version_tags, check_tag
 
 
 class SpecManager:
+    """Class to wrangle app specs."""
+
     __instance = None
 
-    app_specs = {}
-    type_specs = {}
+    app_specs: dict[str, Any] = {}
+    type_specs: dict[str, Any] = {}
 
-    def __new__(cls):
+    def __new__(cls: type["SpecManager"]) -> "SpecManager":
+        """Generate a new SpecManager instance."""
         if SpecManager.__instance is None:
             SpecManager.__instance = object.__new__(cls)
             SpecManager.__instance.reload()
         return SpecManager.__instance
 
-    def get_spec(self, app_id, tag="release"):
+    def get_spec(
+        self: "SpecManager", app_id: str, tag: str = "release"
+    ) -> dict[str, Any]:
+        """Fetch the specs for an app."""
         self.check_app(app_id, tag, raise_exception=True)
         return self.app_specs[tag][app_id]
 
-    def get_type_spec(self, type_id, raise_exception=True, allow_module_match=True):
+    def get_type_spec(
+        self: "SpecManager",
+        type_id: str,
+        raise_exception: bool = True,
+        allow_module_match: bool = True,
+    ):
+        """Fetch the specs for a type."""
         # if we can't find a full match for a type, try to match just the module
         if (type_id not in self.type_specs) and allow_module_match:
             type_id = type_id.split(".")[0]
         if (type_id not in self.type_specs) and raise_exception:
-            raise ValueError('Unknown type id "{}"'.format(type_id))
+            err_msg = f'Unknown type id "{type_id}"'
```

Comment:
convert `format` to f-string; move error message construction to a variable as it gets printed out twice if you construct the message within the arguments to the error.